web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php

```diff
@@ -95,36 +119,75 @@ public function build() {
 
   /**
    * Returns current weather in user location.
+   *
+   * @return array
+   *   Return weather data in array.
+   *   Default array structure
+   *     [
+   *       'img' => string - URL to icon,
+   *       'temp_c' => float - temparature,
+   *       'location' => string - Location, ex: 'Lutsk, Ukraine',
+   *     ].
+   *   Or return an empty array when got exception.
    */
   private function getWeather() {
     $location = $this->getLocation();
-    $token = $this->config->get('weather.weather_api_token') ?? 'token_not_specified';
 
-    $response = $this->client->get(
-      "http://api.weatherapi.com/v1/current.json?key=$token&q=$location&aqi=no",
-      ['http_errors' => FALSE]
-    );
-
-    if ($response->getStatusCode() == 200) {
-      $data = json_decode($response->getBody()->getContents(), TRUE);
+    if ($cache = $this->getWeatherCache($location)) {
+      $data = $cache->data;
 
       return [
-        'img' => 'http:' . $data['current']['condition']['icon'],
-        'temp_c' => $data['current']['temp_c'],
+        'img' => $data['icon_url'],
+        'temp_c' => $data['temp_c'],
         'location' => $location,
       ];
     }
+    else {
+      $token = $this->config->get('weather.weather_api_token') ?? 'token_not_specified';
+
+      $response = $this->client->get(
+        "http://api.weatherapi.com/v1/current.json?key=$token&q=$location&aqi=no",
+        ['http_errors' => FALSE]
+      );
```

Comment:
We shouldn't even try to get the data if the token is not specified, it does not make sense.